File: scripts/test-geotiff-georef.py
#!/usr/bin/env python3
"""
Test GeoTIFF Georeferencing - End-to-End

Simulates exactly what our geotiff-writer.js produces:
1. Reads NISAR HDF5 to get actual coordinate boundaries
2. Applies the same pixel-center → pixel-edge correction as main.jsx export handler
3. Applies the same pixelScale calculation as geotiff-writer.js buildCOGFile()
4. Writes a test GeoTIFF using raw TIFF bytes (same logic as our JS writer)
5. Verifies with rasterio that corners and CRS are correct

Usage:
    python3 scripts/test-geotiff-georef.py [path-to-h5-file]
"""

import sys
import os
import struct
import numpy as np
import h5py

try:
    import rasterio
    from rasterio.crs import CRS
    HAS_RASTERIO = True
except ImportError:
    HAS_RASTERIO = False
    print("WARNING: rasterio not available, will write but cannot verify")

H5_PATH = (
    sys.argv[1] if len(sys.argv) > 1
    else 'test_data/NISAR_L2_PR_GCOV_013_155_D_091_2005_DHDH_A_20251226T231525_20251226T231556_P05006_N_F_J_001.h5'
)

# ---- TIFF constants (matching geotiff-writer.js) ----
TAG_IMAGE_WIDTH = 256
TAG_IMAGE_LENGTH = 257
TAG_BITS_PER_SAMPLE = 258
TAG_COMPRESSION = 259
TAG_PHOTOMETRIC = 262
TAG_STRIP_OFFSETS = 273
TAG_SAMPLES_PER_PIXEL = 277
TAG_ROWS_PER_STRIP = 278
TAG_STRIP_BYTE_COUNTS = 279
TAG_PLANAR_CONFIG = 284
TAG_EXTRA_SAMPLES = 338
TAG_SAMPLE_FORMAT = 339
TAG_MODEL_TIEPOINT = 33922
TAG_MODEL_PIXEL_SCALE = 33550
TAG_GEO_KEY_DIRECTORY = 34735

TYPE_SHORT = 3
TYPE_LONG = 4
TYPE_DOUBLE = 12

KEY_GT_MODEL_TYPE = 1024
KEY_GT_RASTER_TYPE = 1025
KEY_PROJECTED_CS_TYPE = 3072
KEY_GEOGRAPHIC_TYPE = 2048
MODEL_TYPE_PROJECTED = 1
MODEL_TYPE_GEOGRAPHIC = 2
RASTER_TYPE_PIXEL_IS_AREA = 1


def write_minimal_geotiff(path, width, height, bounds, epsg, rgba_data):
    """Write a minimal stripped GeoTIFF matching geotiff-writer.js logic."""
    min_x, min_y, max_x, max_y = bounds

    # This matches geotiff-writer.js buildCOGFile() lines 393-394:
    # pixelScaleX = (maxX - minX) / levels[0].width
    # pixelScaleY = (maxY - minY) / levels[0].height
    pixel_scale_x = (max_x - min_x) / width
    pixel_scale_y = (max_y - min_y) / height

    # Build IFD entries as (tag, type, count, value_or_values)
    entries = []
    entries.append((TAG_IMAGE_WIDTH, TYPE_LONG, 1, [width]))
    entries.append((TAG_IMAGE_LENGTH, TYPE_LONG, 1, [height]))
    entries.append((TAG_BITS_PER_SAMPLE, TYPE_SHORT, 4, [8, 8, 8, 8]))
    entries.append((TAG_COMPRESSION, TYPE_SHORT, 1, [1]))  # No compression for test
    entries.append((TAG_PHOTOMETRIC, TYPE_SHORT, 1, [2]))  # RGB
    entries.append((TAG_STRIP_OFFSETS, TYPE_LONG, 1, [0]))  # placeholder
    entries.append((TAG_SAMPLES_PER_PIXEL, TYPE_SHORT, 1, [4]))
    entries.append((TAG_ROWS_PER_STRIP, TYPE_LONG, 1, [height]))
    entries.append((TAG_STRIP_BYTE_COUNTS, TYPE_LONG, 1, [width * height * 4]))
    entries.append((TAG_PLANAR_CONFIG, TYPE_SHORT, 1, [1]))
    entries.append((TAG_EXTRA_SAMPLES, TYPE_SHORT, 1, [1]))  # Associated alpha
    entries.append((TAG_SAMPLE_FORMAT, TYPE_SHORT, 4, [1, 1, 1, 1]))

    # GeoTIFF tags — matches geotiff-writer.js buildIFD()
    # ModelTiepoint: (0, 0, 0) -> (minX, maxY, 0)
    entries.append((TAG_MODEL_TIEPOINT, TYPE_DOUBLE, 6, [0, 0, 0, min_x, max_y, 0]))
    # ModelPixelScale
    entries.append((TAG_MODEL_PIXEL_SCALE, TYPE_DOUBLE, 3, [pixel_scale_x, pixel_scale_y, 0]))

    # GeoKeyDirectory
    is_geographic = 4000 <= epsg < 5000
    model_type = MODEL_TYPE_GEOGRAPHIC if is_geographic else MODEL_TYPE_PROJECTED
    cs_key = KEY_GEOGRAPHIC_TYPE if is_geographic else KEY_PROJECTED_CS_TYPE
    geo_keys = [
        1, 1, 0, 3,
        KEY_GT_MODEL_TYPE, 0, 1, model_type,
        KEY_GT_RASTER_TYPE, 0, 1, RASTER_TYPE_PIXEL_IS_AREA,
        cs_key, 0, 1, epsg,
    ]
    entries.append((TAG_GEO_KEY_DIRECTORY, TYPE_SHORT, len(geo_keys), geo_keys))

    # Sort by tag
    entries.sort(key=lambda e: e[0])

    # Calculate sizes
    type_sizes = {TYPE_SHORT: 2, TYPE_LONG: 4, TYPE_DOUBLE: 8}
    header_size = 8
    num_entries = len(entries)
    ifd_size = 2 + num_entries * 12 + 4

    # Overflow data (values that don't fit in 4 bytes)
    overflow_entries = []
    overflow_size = 0
    for i, (tag, typ, count, values) in enumerate(entries):
        byte_size = type_sizes[typ] * count
        if byte_size > 4:
            overflow_entries.append(i)
            overflow_size += byte_size
            if overflow_size % 2 != 0:
                overflow_size += 1

    ifd_offset = header_size
    overflow_offset = ifd_offset + ifd_size
    strip_offset = overflow_offset + overflow_size
    strip_size = width * height * 4
    total_size = strip_offset + strip_size

    # Fix strip offset
    for i, (tag, typ, count, values) in enumerate(entries):
        if tag == TAG_STRIP_OFFSETS:
            entries[i] = (tag, typ, count, [strip_offset])

    buf = bytearray(total_size)
    view = memoryview(buf)

    # TIFF header
    struct.pack_into('<2sHI', buf, 0, b'II', 42, ifd_offset)

    # IFD
    pos = ifd_offset
    struct.pack_into('<H', buf, pos, num_entries)
    pos += 2

    cur_overflow = overflow_offset
    for i, (tag, typ, count, values) in enumerate(entries):
        byte_size = type_sizes[typ] * count

        struct.pack_into('<HHI', buf, pos, tag, typ, count)
        pos += 8

        if byte_size > 4:
            # Pointer to overflow
            struct.pack_into('<I', buf, pos, cur_overflow)
            # Write overflow data
            opos = cur_overflow
            for v in values:
                if typ == TYPE_SHORT:
                    struct.pack_into('<H', buf, opos, int(v))
                    opos += 2
                elif typ == TYPE_LONG:
                    struct.pack_into('<I', buf, opos, int(v))
                    opos += 4
                elif typ == TYPE_DOUBLE:
                    struct.pack_into('<d', buf, opos, float(v))
                    opos += 8
            cur_overflow += byte_size
            if cur_overflow % 2 != 0:
                cur_overflow += 1
        else:
            # Inline value
            if count == 1:
                if typ == TYPE_SHORT:
                    struct.pack_into('<H', buf, pos, int(values[0]))
                elif typ == TYPE_LONG:
                    struct.pack_into('<I', buf, pos, int(values[0]))
            else:
                vpos = pos
                for v in values:
                    if typ == TYPE_SHORT:
                        struct.pack_into('<H', buf, vpos, int(v))
                        vpos += 2

        pos += 4

    # Next IFD pointer = 0
    struct.pack_into('<I', buf, pos, 0)

    # Write pixel data
    buf[strip_offset:strip_offset + strip_size] = bytes(rgba_data)

    with open(path, 'wb') as f:
        f.write(buf)

    return pixel_scale_x, pixel_scale_y


def main():
    print(f"\n{'='*60}")
    print(f"  GeoTIFF Georeferencing End-to-End Test")
    print(f"{'='*60}")
    print(f"HDF5 file: {os.path.basename(H5_PATH)}\n")

    # ---- Step 1: Read HDF5 ----
    with h5py.File(H5_PATH, 'r') as f:
        band = None
        for b in ['LSAR', 'SSAR']:
            if f'/science/{b}/GCOV/grids/frequencyA' in f:
                band = b
                break
        assert band, "Could not find LSAR or SSAR"

        freq = 'A'
        base = f'/science/{band}/GCOV/grids/frequency{freq}'

        epsg = int(f[f'{base}/projection'][()])
        x_coords = f[f'{base}/xCoordinates'][:]
        y_coords = f[f'{base}/yCoordinates'][:]
        x_spacing = float(f[f'{base}/xCoordinateSpacing'][()])
        y_spacing = float(f[f'{base}/yCoordinateSpacing'][()])

    width_full = len(x_coords)
    height_full = len(y_coords)

    min_x = min(x_coords[0], x_coords[-1])
    max_x = max(x_coords[0], x_coords[-1])
    min_y = min(y_coords[0], y_coords[-1])
    max_y = max(y_coords[0], y_coords[-1])
    raw_bounds = [min_x, min_y, max_x, max_y]

    print(f"EPSG: {epsg}")
    print(f"Dimensions: {width_full} x {height_full}")
    print(f"Native spacing: {abs(x_spacing):.6f} x {abs(y_spacing):.6f}")
    print(f"Raw bounds (pixel-center): [{min_x:.2f}, {min_y:.2f}, {max_x:.2f}, {max_y:.2f}]")

    # ---- Step 2: Apply pixel-center → pixel-edge correction ----
    # This is what main.jsx export handler now does:
    half_px = abs(x_spacing) / 2
    half_py = abs(y_spacing) / 2
    export_bounds = [
        min_x - half_px,
        min_y - half_py,
        max_x + half_px,
        max_y + half_py,
    ]
    print(f"Export bounds (pixel-edge): [{export_bounds[0]:.2f}, {export_bounds[1]:.2f}, {export_bounds[2]:.2f}, {export_bounds[3]:.2f}]")

    # ---- Step 3: Write test GeoTIFF (simulating geotiff-writer.js) ----
    # Use a small test image but with the FULL bounds (same as real export)
    test_w = 64
    test_h = 64
    test_rgba = bytearray(test_w * test_h * 4)
    for i in range(test_w * test_h):
        test_rgba[i * 4] = 255      # R
        test_rgba[i * 4 + 1] = 0    # G
        test_rgba[i * 4 + 2] = 0    # B
        test_rgba[i * 4 + 3] = 255  # A

    out_path = 'test_data/test_georef_jswriter.tif'
    ps_x, ps_y = write_minimal_geotiff(out_path, test_w, test_h, export_bounds, epsg, test_rgba)

    print(f"\nWrote: {out_path}")
    print(f"  Writer pixel scale: {ps_x:.6f} x {ps_y:.6f}")

    # ---- Step 4: Also test with full-res dimensions ----
    # Simulate the pixel scale calculation with actual export dimensions (e.g., 4×4 multilook)
    for ml in [1, 4, 8]:
        ew = width_full // ml if ml > 1 else min(width_full, 4096)
        eh = height_full // ml if ml > 1 else min(height_full, 4096)
        ps_x_ml = (export_bounds[2] - export_bounds[0]) / ew
        ps_y_ml = (export_bounds[3] - export_bounds[1]) / eh
        expected_ps = abs(x_spacing) * ml
        diff = abs(ps_x_ml - expected_ps)
        status = "PASS" if diff < 0.01 else "FAIL"
        print(f"  [{status}] {ml}×{ml} multilook: {ew}x{eh} → pixel scale {ps_x_ml:.6f}m (expected {expected_ps:.1f}m, diff={diff:.6f})")

    # ---- Step 5: Verify with rasterio ----
    if not HAS_RASTERIO:
        print("\nSkipping rasterio verification (not installed)")
        return

    print(f"\n{'='*60}")
    print(f"  VERIFICATION (rasterio)")
    print(f"{'='*60}")

    all_pass = True

    with rasterio.open(out_path) as src:
        b = src.bounds
        t = src.transform
        res = src.res

        print(f"\nFile: {out_path}")
        print(f"  Dimensions: {src.width} x {src.height}, {src.count} bands")
        print(f"  CRS: {src.crs}")
        print(f"  Origin (UL): ({t.c:.2f}, {t.f:.2f})")
        print(f"  Pixel size: ({t.a:.6f}, {t.e:.6f})")
        print(f"  Bounds: [{b.left:.2f}, {b.bottom:.2f}, {b.right:.2f}, {b.top:.2f}]")

        # Check 1: CRS
        actual_epsg = src.crs.to_epsg() if src.crs else None
        ok = actual_epsg == epsg
        status = "PASS" if ok else "FAIL"
        if not ok: all_pass = False
        print(f"\n  [{status}] EPSG code: {actual_epsg} (expected {epsg})")

        # Check 2: Bounds match pixel-edge
        expected_edge = (export_bounds[0], export_bounds[1], export_bounds[2], export_bounds[3])
        actual_bounds = (b.left, b.bottom, b.right, b.top)
        bounds_match = all(abs(a - e) < 1.0 for a, e in zip(actual_bounds, expected_edge))
        status = "PASS" if bounds_match else "FAIL"
        if not bounds_match: all_pass = False
        print(f"  [{status}] Bounds match pixel-EDGE")
        if not bounds_match:
            print(f"    Actual:   [{b.left:.2f}, {b.bottom:.2f}, {b.right:.2f}, {b.top:.2f}]")
            print(f"    Expected: [{expected_edge[0]:.2f}, {expected_edge[1]:.2f}, {expected_edge[2]:.2f}, {expected_edge[3]:.2f}]")

        # Check 3: Bounds do NOT match pixel-center (they shouldn't)
        expected_center = (raw_bounds[0], raw_bounds[1], raw_bounds[2], raw_bounds[3])
        center_match = all(abs(a - e) < 1.0 for a, e in zip(actual_bounds, expected_center))
        status = "PASS" if not center_match else "FAIL"
        if center_match: all_pass = False
        print(f"  [{status}] Bounds differ from pixel-CENTER (should differ by ±{half_px:.0f}m)")

        # Check 4: Origin is at pixel edge upper-left
        origin_ok = abs(t.c - export_bounds[0]) < 0.01 and abs(t.f - export_bounds[3]) < 0.01
        status = "PASS" if origin_ok else "FAIL"
        if not origin_ok: all_pass = False
        print(f"  [{status}] Origin at pixel-edge UL: ({t.c:.2f}, {t.f:.2f})")

        # Check 5: Y pixel scale is negative (north-up)
        y_neg = t.e < 0
        status = "PASS" if y_neg else "FAIL"
        if not y_neg: all_pass = False
        print(f"  [{status}] Y pixel scale is negative (north-up): {t.e:.6f}")

    print(f"\n{'='*60}")
    if all_pass:
        print(f"  ALL TESTS PASSED")
    else:
        print(f"  SOME TESTS FAILED")
    print(f"{'='*60}\n")

    return all_pass


if __name__ == '__main__':
    success = main()
    sys.exit(0 if success else 1)

File: scripts/test-geotiff-georef.mjs
/**
 * Test GeoTIFF Georeferencing
 *
 * Reads a NISAR HDF5 file to extract the actual coordinate boundaries,
 * then writes a small test GeoTIFF using our geotiff-writer and saves it
 * for verification with GDAL (see verify-georef.py).
 *
 * Usage:
 *   node scripts/test-geotiff-georef.mjs [path-to-h5-file]
 */

import { readFileSync, writeFileSync } from 'fs';
import { join, basename } from 'path';
import h5wasm from 'h5wasm';

// We can't import the ES module geotiff-writer directly in Node,
// so we replicate the minimal GeoTIFF writing logic here for the test.

const TILE_SIZE = 512;

// TIFF tag IDs
const TAG_NEW_SUBFILE_TYPE = 254;
const TAG_IMAGE_WIDTH = 256;
const TAG_IMAGE_LENGTH = 257;
const TAG_BITS_PER_SAMPLE = 258;
const TAG_COMPRESSION = 259;
const TAG_PHOTOMETRIC = 262;
const TAG_SAMPLES_PER_PIXEL = 277;
const TAG_PLANAR_CONFIG = 284;
const TAG_TILE_WIDTH = 322;
const TAG_TILE_LENGTH = 323;
const TAG_TILE_OFFSETS = 324;
const TAG_TILE_BYTE_COUNTS = 325;
const TAG_EXTRA_SAMPLES = 338;
const TAG_SAMPLE_FORMAT = 339;
const TAG_MODEL_TIEPOINT = 33922;
const TAG_MODEL_PIXEL_SCALE = 33550;
const TAG_GEO_KEY_DIRECTORY = 34735;

// TIFF type IDs
const TYPE_SHORT = 3;
const TYPE_LONG = 4;
const TYPE_DOUBLE = 12;

// GeoKey constants
const KEY_GT_MODEL_TYPE = 1024;
const KEY_GT_RASTER_TYPE = 1025;
const KEY_PROJECTED_CS_TYPE = 3072;
const KEY_GEOGRAPHIC_TYPE = 2048;
const MODEL_TYPE_PROJECTED = 1;
const MODEL_TYPE_GEOGRAPHIC = 2;
const RASTER_TYPE_PIXEL_IS_AREA = 1;

// NISAR HDF5 path builder
function nisarPaths(band = 'LSAR', productType = 'GCOV') {
  const base = `/science/${band}/${productType}`;
  return {
    projection: (f) => `${base}/grids/frequency${f}/projection`,
    xCoordinates: (f) => `${base}/grids/frequency${f}/xCoordinates`,
    yCoordinates: (f) => `${base}/grids/frequency${f}/yCoordinates`,
    xCoordinateSpacing: (f) => `${base}/grids/frequency${f}/xCoordinateSpacing`,
    yCoordinateSpacing: (f) => `${base}/grids/frequency${f}/yCoordinateSpacing`,
  };
}

function safeGet(h5file, path) {
  try { return h5file.get(path); } catch { return null; }
}

async function main() {
  const h5Path = process.argv[2] ||
    'test_data/NISAR_L2_PR_GCOV_013_155_D_091_2005_DHDH_A_20251226T231525_20251226T231556_P05006_N_F_J_001.h5';

  console.log(`\n=== GeoTIFF Georeferencing Test ===`);
  console.log(`HDF5 file: ${h5Path}\n`);

  // Initialize h5wasm
  await h5wasm.ready;
  const h5file = new h5wasm.File(h5Path, 'r');

  // Try LSAR and SSAR
  let band = 'LSAR';
  let frequency = 'A';
  let paths = nisarPaths(band);

  // Check if LSAR exists, otherwise try SSAR
  let projDs = safeGet(h5file, paths.projection(frequency));
  if (!projDs) {
    band = 'SSAR';
    paths = nisarPaths(band);
    projDs = safeGet(h5file, paths.projection(frequency));
  }

  // === 1. Extract EPSG code ===
  let epsgCode = 4326;
  if (projDs) {
    const projVal = projDs.value;
    if (typeof projVal === 'number' && projVal > 0) {
      epsgCode = projVal;
    }
  }
  console.log(`EPSG code: ${epsgCode}`);

  // === 2. Extract coordinate arrays ===
  const xCoordsDs = safeGet(h5file, paths.xCoordinates(frequency));
  const yCoordsDs = safeGet(h5file, paths.yCoordinates(frequency));

  if (!xCoordsDs || !yCoordsDs) {
    console.error('ERROR: Could not find xCoordinates/yCoordinates datasets');
    process.exit(1);
  }

  const xCoords = xCoordsDs.value;
  const yCoords = yCoordsDs.value;

  console.log(`\nCoordinate arrays:`);
  console.log(`  xCoordinates: ${xCoords.length} values`);
  console.log(`    first: ${xCoords[0]}`);
  console.log(`    last:  ${xCoords[xCoords.length - 1]}`);
  console.log(`  yCoordinates: ${yCoords.length} values`);
  console.log(`    first: ${yCoords[0]}`);
  console.log(`    last:  ${yCoords[yCoords.length - 1]}`);

  // === 3. Calculate bounds ===
  const minX = Math.min(xCoords[0], xCoords[xCoords.length - 1]);
  const maxX = Math.max(xCoords[0], xCoords[xCoords.length - 1]);
  const minY = Math.min(yCoords[0], yCoords[yCoords.length - 1]);
  const maxY = Math.max(yCoords[0], yCoords[yCoords.length - 1]);
  const bounds = [minX, minY, maxX, maxY];

  const width = xCoords.length;
  const height = yCoords.length;

  console.log(`\nImage dimensions: ${width} x ${height}`);
  console.log(`Bounds [minX, minY, maxX, maxY]: [${bounds.join(', ')}]`);

  // === 4. Pixel spacing ===
  const xSpacingDs = safeGet(h5file, paths.xCoordinateSpacing(frequency));
  const ySpacingDs = safeGet(h5file, paths.yCoordinateSpacing(frequency));
  const xSpacing = xSpacingDs ? xSpacingDs.value : null;
  const ySpacing = ySpacingDs ? ySpacingDs.value : null;

  console.log(`\nNative pixel spacing from HDF5:`);
  console.log(`  xCoordinateSpacing: ${xSpacing}`);
  console.log(`  yCoordinateSpacing: ${ySpacing}`);

  const calcPixelSizeX = (maxX - minX) / (width - 1);
  const calcPixelSizeY = (maxY - minY) / (height - 1);
  console.log(`Calculated pixel size from bounds/(N-1):`);
  console.log(`  pixelSizeX: ${calcPixelSizeX.toFixed(6)}`);
  console.log(`  pixelSizeY: ${calcPixelSizeY.toFixed(6)}`);

  // === 5. What geotiff-writer uses ===
  // geotiff-writer.js line 393-394: pixelScale = (max - min) / width (NOT width-1)
  const writerPixelScaleX = (maxX - minX) / width;
  const writerPixelScaleY = (maxY - minY) / height;
  console.log(`\ngeotiff-writer pixel scale (bounds/dimension):`);
  console.log(`  pixelScaleX: ${writerPixelScaleX.toFixed(6)}`);
  console.log(`  pixelScaleY: ${writerPixelScaleY.toFixed(6)}`);

  // === 6. Expected corner coordinates ===
  console.log(`\n--- Expected GeoTIFF corners ---`);
  console.log(`  Upper-left:  (${minX}, ${maxY})`);
  console.log(`  Upper-right: (${maxX}, ${maxY})`);
  console.log(`  Lower-left:  (${minX}, ${minY})`);
  console.log(`  Lower-right: (${maxX}, ${minY})`);

  // === 7. Write a small test GeoTIFF ===
  // Use a 64x64 pixel test image (solid red with alpha)
  const testWidth = 64;
  const testHeight = 64;
  const testData = new Uint8Array(testWidth * testHeight * 4);
  for (let i = 0; i < testWidth * testHeight; i++) {
    testData[i * 4] = 255;     // R
    testData[i * 4 + 1] = 0;   // G
    testData[i * 4 + 2] = 0;   // B
    testData[i * 4 + 3] = 255; // A
  }

  // Write a minimal (non-COG) GeoTIFF for testing
  const outPath = 'test_data/test_georef.tif';
  const buffer = writeMinimalGeoTIFF(testData, testWidth, testHeight, bounds, epsgCode);
  writeFileSync(outPath, Buffer.from(buffer));

  console.log(`\nWrote test GeoTIFF: ${outPath}`);
  console.log(`  Dimensions: ${testWidth}x${testHeight}`);
  console.log(`  Bounds: [${bounds.join(', ')}]`);
  console.log(`  EPSG: ${epsgCode}`);
  console.log(`  Pixel scale: ${((maxX - minX) / testWidth).toFixed(2)} x ${((maxY - minY) / testHeight).toFixed(2)}`);
  console.log(`\nRun: python scripts/verify-georef.py ${outPath}`);

  h5file.close();
}

/**
 * Write a minimal stripped (non-tiled) GeoTIFF for test purposes.
 * Simpler than the COG writer - just enough to test georeferencing tags.
 */
function writeMinimalGeoTIFF(rgbaData, width, height, bounds, epsgCode) {
  const [minX, minY, maxX, maxY] = bounds;
  const pixelScaleX = (maxX - minX) / width;
  const pixelScaleY = (maxY - minY) / height;

  // Build IFD entries
  const entries = [];

  entries.push({ tag: TAG_IMAGE_WIDTH, type: TYPE_LONG, count: 1, value: width });
  entries.push({ tag: TAG_IMAGE_LENGTH, type: TYPE_LONG, count: 1, value: height });
  entries.push({ tag: TAG_BITS_PER_SAMPLE, type: TYPE_SHORT, count: 4, values: [8, 8, 8, 8] });
  entries.push({ tag: TAG_COMPRESSION, type: TYPE_SHORT, count: 1, value: 1 }); // No compression
  entries.push({ tag: TAG_PHOTOMETRIC, type: TYPE_SHORT, count: 1, value: 2 }); // RGB
  entries.push({ tag: TAG_SAMPLES_PER_PIXEL, type: TYPE_SHORT, count: 1, value: 4 });
  entries.push({ tag: TAG_PLANAR_CONFIG, type: TYPE_SHORT, count: 1, value: 1 }); // Chunky

  // StripOffsets - placeholder
  entries.push({ tag: 273, type: TYPE_LONG, count: 1, value: 0 }); // StripOffsets
  entries.push({ tag: 278, type: TYPE_LONG, count: 1, value: height }); // RowsPerStrip
  entries.push({ tag: 279, type: TYPE_LONG, count: 1, value: width * height * 4 }); // StripByteCounts

  entries.push({ tag: TAG_EXTRA_SAMPLES, type: TYPE_SHORT, count: 1, value: 1 }); // Associated alpha
  entries.push({ tag: TAG_SAMPLE_FORMAT, type: TYPE_SHORT, count: 4, values: [1, 1, 1, 1] });

  // GeoTIFF tags
  entries.push({ tag: TAG_MODEL_TIEPOINT, type: TYPE_DOUBLE, count: 6, values: [0, 0, 0, minX, maxY, 0] });
  entries.push({ tag: TAG_MODEL_PIXEL_SCALE, type: TYPE_DOUBLE, count: 3, values: [pixelScaleX, pixelScaleY, 0] });

  const isGeographic = epsgCode >= 4000 && epsgCode < 5000;
  const modelType = isGeographic ? MODEL_TYPE_GEOGRAPHIC : MODEL_TYPE_PROJECTED;
  const csKeyId = isGeographic ? KEY_GEOGRAPHIC_TYPE : KEY_PROJECTED_CS_TYPE;

  entries.push({
    tag: TAG_GEO_KEY_DIRECTORY, type: TYPE_SHORT, count: 16, values: [
      1, 1, 0, 3,
      KEY_GT_MODEL_TYPE, 0, 1, modelType,
      KEY_GT_RASTER_TYPE, 0, 1, RASTER_TYPE_PIXEL_IS_AREA,
      csKeyId, 0, 1, epsgCode,
    ]
  });

  // Sort entries by tag
  entries.sort((a, b) => a.tag - b.tag);

  // Calculate sizes
  const numEntries = entries.length;
  const headerSize = 8;
  const ifdSize = 2 + numEntries * 12 + 4;

  // Calculate overflow data sizes
  let overflowSize = 0;
  for (const entry of entries) {
    const byteSize = getByteSize(entry);
    if (byteSize > 4) {
      entry.needsOverflow = true;
      entry.overflowBytes = byteSize;
      overflowSize += byteSize;
      if (overflowSize % 2 !== 0) overflowSize++;
    }
  }

  const ifdOffset = headerSize;
  const overflowOffset = ifdOffset + ifdSize;
  const stripOffset = overflowOffset + overflowSize;
  const stripSize = width * height * 4;
  const totalSize = stripOffset + stripSize;

  // Allocate buffer
  const buffer = new ArrayBuffer(totalSize);
  const view = new DataView(buffer);
  const bytes = new Uint8Array(buffer);

  // Write TIFF header (Little-endian)
  view.setUint16(0, 0x4949, true); // II = Little endian
  view.setUint16(2, 42, true);     // Magic
  view.setUint32(4, ifdOffset, true); // First IFD offset

  // Write IFD
  let pos = ifdOffset;
  view.setUint16(pos, numEntries, true); pos += 2;

  // Fix StripOffsets value
  for (const entry of entries) {
    if (entry.tag === 273) entry.value = stripOffset;
  }

  let curOverflow = overflowOffset;

  for (const entry of entries) {
    view.setUint16(pos, entry.tag, true); pos += 2;
    view.setUint16(pos, entry.type, true); pos += 2;
    view.setUint32(pos, entry.count, true); pos += 4;

    if (entry.needsOverflow) {
      view.setUint32(pos, curOverflow, true); pos += 4;
      // Write overflow data
      writeValues(view, curOverflow, entry);
      curOverflow += entry.overflowBytes;
      if (curOverflow % 2 !== 0) curOverflow++;
    } else if (entry.count === 1) {
      // Single value fits in 4 bytes
      if (entry.type === TYPE_SHORT) {
        view.setUint16(pos, entry.value, true);
        pos += 4;
      } else if (entry.type === TYPE_LONG) {
        view.setUint32(pos, entry.value, true);
        pos += 4;
      }
    } else {
      // Multiple small values fit in 4 bytes
      let vPos = pos;
      for (const v of entry.values) {
        if (entry.type === TYPE_SHORT) {
          view.setUint16(vPos, v, true);
          vPos += 2;
        }
      }
      pos += 4;
    }
  }

  // Next IFD pointer (0 = no more IFDs)
  view.setUint32(pos, 0, true);

  // Write strip data (RGBA pixels)
  bytes.set(rgbaData, stripOffset);

  return buffer;
}

function getByteSize(entry) {
  const typeSize = { [TYPE_SHORT]: 2, [TYPE_LONG]: 4, [TYPE_DOUBLE]: 8 };
  return (typeSize[entry.type] || 1) * entry.count;
}

function writeValues(view, offset, entry) {
  let pos = offset;
  const values = entry.values || [entry.value];
  for (const v of values) {
    if (entry.type === TYPE_SHORT) {
      view.setUint16(pos, v, true); pos += 2;
    } else if (entry.type === TYPE_LONG) {
      view.setUint32(pos, v, true); pos += 4;
    } else if (entry.type === TYPE_DOUBLE) {
      view.setFloat64(pos, v, true); pos += 8;
    }
  }
}

main().catch(e => {
  console.error('Error:', e);
  process.exit(1);
});
